### PR TITLE
[SPARK-21165] [SQL] [2.2] Use executedPlan instead of analyzedPlan in INSERT AS SELECT

### PR DESCRIPTION
**Pull request by gatorsmile**
### What changes were proposed in this pull request?

The input query schema of INSERT AS SELECT could be changed after optimization. For example, the following query's output schema is changed by the rule `SimplifyCasts` and `RemoveRedundantAliases`.
```SQL
 SELECT word, length, cast(first as string) as first FROM view1
```

This PR is to fix the issue in Spark 2.2. Instead of using the analyzed plan of the input query, this PR use its executed plan to determine the attributes in `FileFormatWriter`.

The related issue in the master branch has been fixed by https://github.com/apache/spark/pull/18064. After this PR is merged, I will submit a separate PR to merge the test case to the master.

### How was this patch tested?
Added a test case

### CODE REVIEW

**Comment by gatorsmile** [comment]
cc @cloud-fan

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala [hunk 1]
```diff
@@ -111,9 +111,18 @@ object FileFormatWriter extends Logging {
...
+    val partitionColumns = partitionColumnNames.map { col =>
+      val nameEquality = sparkSession.sessionState.conf.resolver
+      allColumns.find(f => nameEquality(f.name, col)).getOrElse {
+        throw new RuntimeException(
+          s"Partition column $col not found in schema ${queryExecution.executedPlan.schema}")
+      }
+    }
     val partitionSet = AttributeSet(partitionColumns)
-    val dataColumns = queryExecution.logical.output.filterNot(partitionSet.contains)
+    val dataColumns = queryExecution.executedPlan.output.filterNot(partitionSet.contains)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala]: nit: use `allColumns` here

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
LGTM, pending jenkins

**Comment by cloud-fan** [comment]
hmm, seems a legitimate test failure...

**Comment by jiangxb1987** [comment]
The jenkins has been unstable recently.

**Comment by jiangxb1987** [comment]
retest this please

**Comment by gatorsmile** [comment]
Weird. The test case passed in my local environment. Need to do more investigation.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala [hunk 2]
```diff
@@ -127,11 +127,11 @@ case class PreprocessTableCreation(sparkSession: SparkSession) extends Rule[Logi
...
 
-      // As we are inserting into an existing table, we should respect the existing schema and
-      // adjust the column order of the given dataframe according to it, or throw exception
-      // if the column names do not match.
+      // As we are inserting into an existing table, we should respect the existing schema, preserve
+      // the case and adjust the column order of the given DataFrame according to it, or throw
+      // an exception if the column names do not match.
       val adjustedColumns = tableCols.map { col =>
-        query.resolve(Seq(col), resolver).getOrElse {
+        query.resolve(Seq(col), resolver).map(Alias(_, col)()).getOrElse {
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala]: Need to add an alias for enforcing the query to preserve the original name of table schema, whose case could be different from the underlying query schema

Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala]: ah good catch!

**Comment by gatorsmile** [comment]
retest this please

**Comment by cloud-fan** [comment]
thanks, merging to 2.2!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala [hunk 3]
```diff
@@ -111,9 +111,18 @@ object FileFormatWriter extends Logging {
     job.setOutputValueClass(classOf[InternalRow])
     FileOutputFormat.setOutputPath(job, new Path(outputSpec.outputPath))
 
-    val allColumns = queryExecution.logical.output
+    val allColumns = queryExecution.executedPlan.output
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala]: This is problematic.  The physical plan may have different schema from logical plan(schema name may be different), and the writer should respect the logical schema as that what users expects.

Comment (gatorsmile) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala]: Yes. We should always use `analyzed.output`

**Pull request closed without merging**